[pytorch] codegen flags to whitelist op registrations / generate to separate files

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32451 [pytorch] codegen flags to whitelist op registrations / generate to separate files**
* #32275 [pytorch] move type_derived_methods out of anonymous namespace

This PR adds a few new parameters to ATen codegen script:

```
1. op_registration_whitelist
Can be used to filter op registrations for selective build;

2. type_whitelist
Can be used to filter types (CPUType, CUDAType, ...) for selective build;

3. per_op_registration
When set it will group function registrations by op name and write to separate files;
```

1 & 2 are introduced for mobile custom build without relying on static dispatch;
3 is introduced to solve custom build with multi-library / multi-model (needed by FB
internal build - see more details: https://fb.quip.com/ZVh1AgOKW8Vv).

These flags should work independently with each other (and independent to USE_STATIC_DISPATCH).

Not setting them should have no effect compared to master.

Differential Revision: [D19427919](https://our.internmc.facebook.com/intern/diff/D19427919/)

ljk53 committed:
[pytorch] codegen flags to whitelist op registrations / generate to separate files

This PR adds a few new parameters to ATen codegen script:

```
1. op_registration_whitelist
Can be used to filter op registrations for selective build;

2. type_whitelist
Can be used to filter types (CPUType, CUDAType, ...) for selective build;

3. per_op_registration
When set it will group function registrations by op name and write to separate files;
```

1 & 2 are introduced for mobile custom build without relying on static dispatch;
3 is introduced to solve custom build with multi-library / multi-model (needed by FB
internal build - see more details: https://fb.quip.com/ZVh1AgOKW8Vv).

These flags should work independently with each other (and independent to USE_STATIC_DISPATCH).

Not setting them should have no effect compared to master.

Differential Revision: [D19427919](https://our.internmc.facebook.com/intern/diff/D19427919/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit c01b056f:


* **1/3** failures introduced in this PR
* **2/3** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c01b056fc043d6368950290d5e60d9f1397c8ac3).

### :snowflake: 2 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4307050) pytorch_linux_xenial_py2_7_9_build (1/2)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=88968671) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=88968671)) :snowflake:
<details>
<summary>
<code>E: Failed to fetch https://nvidia.github.io/libnvidia-container/ubuntu16.04/amd64/Packages  Writing more data than expected (55935 > 51667)</code>
</summary>

```
96% [Waiting for headers] 96% [40 Translation-en store 0 B] [Waiting for headers]                                                         Get:53 http://archive.ubuntu.com/ubuntu xenial-updates/multiverse amd64 Packages [16.8 kB] 
                                                                    96% [40 Translation-en store 0 B]                                   Get:54 http://archive.ubuntu.com/ubuntu xenial-updates/multiverse Translation-en [8,468 B] 
                                                                           96% [40 Translation-en store 0 B]                                   Get:55 http://archive.ubuntu.com/ubuntu xenial-backports/main Sources [4,848 B] 
                                                                    96% [40 Translation-en store 0 B]                                   Get:56 http://archive.ubuntu.com/ubuntu xenial-backports/universe Sources [7,120 B] 
                                                                    96% [40 Translation-en store 0 B]                                   Get:57 http://archive.ubuntu.com/ubuntu xenial-backports/main amd64 Packages [7,280 B] 
                                                                     96% [40 Translation-en store 0 B]                                   Get:58 http://archive.ubuntu.com/ubuntu xenial-backports/main Translation-en [4,456 B] 
                                                                           96% [40 Translation-en store 0 B]                                   Get:59 http://archive.ubuntu.com/ubuntu xenial-backports/universe amd64 Packages [8,064 B] 
                                                                     96% [40 Translation-en store 0 B]                                   Get:60 http://archive.ubuntu.com/ubuntu xenial-backports/universe Translation-en [4,328 B] 
                                   100% [Working]                Fetched 28.8 MB in 5s (4,887 kB/s) 
Reading package lists... 99%  Reading package lists... Done  
E: Failed to fetch https://nvidia.github.io/libnvidia-container/ubuntu16.04/amd64/Packages  Writing more data than expected (55935 > 51667) 
E: Some index files failed to download. They have been ignored, or old ones used instead. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4307063) pytorch_linux_xenial_py3_6_gcc5_4_build (2/2)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=88968739) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=88968739)) :snowflake:
<details>
<summary>
<code>E: Failed to fetch https://nvidia.github.io/libnvidia-container/ubuntu16.04/amd64/Packages  Writing more data than expected (55935 > 51667)</code>
</summary>

```
                                                                        96% [40 Translation-en store 0 B]                                   Get:53 http://archive.ubuntu.com/ubuntu xenial-updates/multiverse amd64 Packages [16.8 kB] 
                                                                    96% [40 Translation-en store 0 B]                                   Get:54 http://archive.ubuntu.com/ubuntu xenial-updates/multiverse Translation-en [8,468 B] 
                                                                           96% [40 Translation-en store 0 B]                                   Get:55 http://archive.ubuntu.com/ubuntu xenial-backports/main Sources [4,848 B] 
                                                                    96% [40 Translation-en store 0 B]                                   Get:56 http://archive.ubuntu.com/ubuntu xenial-backports/universe Sources [7,120 B] 
                                                                    96% [40 Translation-en store 0 B]                                   Get:57 http://archive.ubuntu.com/ubuntu xenial-backports/main amd64 Packages [7,280 B] 
                                                                     96% [40 Translation-en store 0 B]                                   Get:58 http://archive.ubuntu.com/ubuntu xenial-backports/main Translation-en [4,456 B] 
96% [Waiting for headers] 96% [42 Translation-en store 0 B] [Waiting for headers]                                                         Get:59 http://archive.ubuntu.com/ubuntu xenial-backports/universe amd64 Packages [8,064 B] 
97% [Waiting for headers] 97% [45 Sources store 0 B] [Waiting for headers]                                                  Get:60 http://archive.ubuntu.com/ubuntu xenial-backports/universe Translation-en [4,328 B] 
                                   100% [Working]                Fetched 28.8 MB in 5s (5,042 kB/s) 
Reading package lists... 99%  Reading package lists... Done  
E: Failed to fetch https://nvidia.github.io/libnvidia-container/ubuntu16.04/amd64/Packages  Writing more data than expected (55935 > 51667) 
E: Some index files failed to download. They have been ignored, or old ones used instead. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -696,7 +698,7 @@ def to_return_type(arg, option):
 
 def create_generic(top_env, declarations):
-    # type: (TopEnvironment, List[FunctionOption]) -> List[OutputDeclaration]
+    # type: (TopEnvironment, List[FunctionOption]) -> Tuple[List[OutputDeclaration], List[Tuple[str, str]]]
```
`Tuple[str, str]` is not a very descriptive type; it would be helpful to know what exactly is the extra stuff we're returning here.

ljk53 replied:
Sure, I'll change this to a NamedTuple.

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -43,6 +43,18 @@
+parser.add_argument(
+    '--op_registration_whitelist',
+    nargs='*',
+    help='filter op registrations by the whitelist (if set)')
```
It would be very helpful to give an example valid entry for the whitelist. Do we whitelist full names with overloads `add.Tensor`? Or just `add`?

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -43,6 +43,18 @@
+parser.add_argument(
+    '--type_whitelist',
+    nargs='*',
+    help='filter types (CPUType, CUDAType, ...) by the whitelist (if set)')
```
Noooo more public references to CPUType XD (no action needed)

gchanan replied:
strongly agree.  CPUType is an implementation detail.  What do you actually need to whitelist?  If it's just stuff from native_functions, just use the keys from there.

ljk53 replied:
> strongly agree. CPUType is an implementation detail. What do you actually need to whitelist? If it's just stuff from native_functions, just use the keys from there.

To clarify, do you mean using CPU / CUDA / MkldnnCPU / QuantizedCPU - dispatch keys used in native_functions.yaml? I'll update.

ezyang replied:
Yep!

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -360,6 +403,28 @@ def filter_by_extension(files, *extensions):
+
+    # Ensure all whitelisted operators have a corresponding registration file.
+    # Generate an empty placeholder file for nonexistent operators, which might
+    # be registered manually instead of via codegen.
```
Hmm, why is the empty placeholder file needed?

ljk53 replied:
It can simplify the BUCK side logic. We can unconditionally create BUCK targets for each operator (with srcs = [ pt_op_register_XXX.cpp ]), and unconditionally add them to the `deps` of feature targets. I'm not sure if there is easy way to make BUCK tolerant targets with nonexistent src files.

You may want to ask why cannot we ignore manually registered ops in the first place (before calling the script)?

A model might be calling regular autogenerated aten ops and/or manually registered ops (e.g. quantized ops), and these ops might be calling ops in the other category. For example:

- codegen registered ops calling manually registered ops:
aten::quantized_lstm -> quantized::linear_dynamic

- manually registered ops calling codegen registered ops:
quantized::relu6 -> aten::empty

Even though we cannot turn on/off the registration of quantized ops via codegen, we need take them into consideration when calculating what codegen registered ops need to be kept for custom build.

We could create some rule / metadata (e.g. `manual_kernel_registration` which doesn't include quantized ops thus incomplete) to handle this, but I feel an empty placeholder is probably the simplest solution.

ezyang replied:
OK, makes sense to me. A good comment to put in the code so we remember it in the future :)

ljk53 replied:
Added more comments.

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -360,6 +403,28 @@ def filter_by_extension(files, *extensions):
+            per_op_registrations[whitelisted_op] = []
+
+    for opname, function_registrations in per_op_registrations.items():
+        fname = 'pt_op_register_%s.cpp' % opname.replace(':', '_')
```
nit: use modern `.format()` interpolation here

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -360,6 +403,28 @@ def filter_by_extension(files, *extensions):
+    for opname, function_registrations in per_op_registrations.items():
+        fname = 'pt_op_register_%s.cpp' % opname.replace(':', '_')
+        # output_dependencies doesn't seem to be used, otherwise it is probably
+        # too late to declare output here?
```
No it's not. The way that `output_dependencies` works is we do a dry run of what files we would write, but don't actually write any of them (that's what file manager manages: it will no-op if we're doing a dry run) so that the build system can account for it. So ideally, you'd have modified `declare_outputs` to have a new logic for computing what per op files you are going to generate, and then `generate_outputs` to actually do the generation. I'm not sure if you can conveniently do this with per-file op registration as I'm not sure if you have enough of the information in `declare_outputs` to say what files will be generated.

ezyang replied:
Maybe, every file has to be specified in the whitelist command line? Then you can populate the set of written out files this way.

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -229,7 +229,9 @@ def TypedDict(name, attrs, total=True):  # type: ignore
+OPERATOR_NAME = CodeTemplate("aten::${operator_name}")
+
+OPERATOR_NAME_FULL = CodeTemplate("""\
     {"aten::${operator_name}", "${overload_name}"},
```
So if I understand correctly, per-op registration doesn't give a separate file to each overload?

ljk53 replied:
Yes, the current granularity is at operator level. The code analyzer doesn't yet extract overload name from op registration/invocation. We could break it down if overload name is widely used and has meaningful impact on custom build size.

ezyang replied:
That's fine, it just wasn't obvious to me that this is what was happening looking at the diff.

ezyang commented:
The basic structure of the PR looks reasonable, but zdevito will have a better sense if there are any other structural things to watch out for.

gchanan commented:
can you list a reasonable invocation (a fake one is okay, one you use is better) so I can run this and see the output?

Also CC @bhosmer.

ljk53 commented:
> can you list a reasonable invocation (a fake one is okay, one you use is better) so I can run this and see the output?
> 
> Also CC @bhosmer.

Option 1: run it directly with arbitrary list of ops, e.g.:
```
python aten/src/ATen/gen.py \
aten/src/ATen/Declarations.cwrap \
aten/src/THNN/generic/THNN.h \
aten/src/THCUNN/generic/THCUNN.h \
aten/src/ATen/nn.yaml \
aten/src/ATen/native/native_functions.yaml \
--source-path aten/src/ATen \
--install_dir /tmp/test \
--per_op_registration \
--type_whitelist CPUType QuantizedCPUType \
--op_registration_whitelist aten::empty aten::fill_ aten::narrow aten::conv2d
```

Option 2: patch D19427938 and follow the instructions in the diff summary to see how it works end-to-end in internal xplat/buck build environment.

ljk53 committed:
Update on "[pytorch] codegen flags to whitelist op registrations / generate to separate files"


This PR adds a few new parameters to ATen codegen script:

```
1. op_registration_whitelist
Can be used to filter op registrations for selective build;

2. type_whitelist
Can be used to filter types (CPUType, CUDAType, ...) for selective build;

3. per_op_registration
When set it will group function registrations by op name and write to separate files;
```

1 & 2 are introduced for mobile custom build without relying on static dispatch;
3 is introduced to solve custom build with multi-library / multi-model (needed by FB
internal build - see more details: https://fb.quip.com/ZVh1AgOKW8Vv).

These flags should work independently with each other (and independent to USE_STATIC_DISPATCH).

Not setting them should have no effect compared to master.

Differential Revision: [D19427919](https://our.internmc.facebook.com/intern/diff/D19427919/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -217,13 +241,33 @@ def format_yaml(data):
+            continue
+        # per op registration
+        try:
+            per_op_registrations[opname].append(registration)
```
Seems like a good use case for defaultdict :>

ljk53 replied:
updated

ezyang reviewed (approved):
okey dokey

zdevito reviewed (commented):
Deferring to Ed here, seems fine as as way to parameterize what is emitted.

ljk53 committed:
Update on "[pytorch] codegen flags to whitelist op registrations / generate to separate files"


This PR adds a few new parameters to ATen codegen script:

```
1. op_registration_whitelist
Can be used to filter op registrations for selective build;

2. type_whitelist
Can be used to filter types (CPUType, CUDAType, ...) for selective build;

3. per_op_registration
When set it will group function registrations by op name and write to separate files;
```

1 & 2 are introduced for mobile custom build without relying on static dispatch;
3 is introduced to solve custom build with multi-library / multi-model (needed by FB
internal build - see more details: https://fb.quip.com/ZVh1AgOKW8Vv).

These flags should work independently with each other (and independent to USE_STATIC_DISPATCH).

Not setting them should have no effect compared to master.

Differential Revision: [D19427919](https://our.internmc.facebook.com/intern/diff/D19427919/)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/gen.py:
```diff
@@ -156,6 +175,8 @@ def backend_to_devicetype(backend):
 ]
 
+# whitelist used to filter op registrations for custom build
+op_registration_whitelist = set(options.op_registration_whitelist or [])
```
nit: semantically, an empty white list would be more consistent if it meant to have **no** operators present, not **all** operators like you're using it here. I think it helps readability if `op_registration_whitelist` doesn't use `set([])` but `None` to disable the whitelist.

ljk53 replied:
Yes I agree - thought about it but didn't know how to write it as short as the current version :p
I'll update.

ljk53 replied:
changed to None

smessmer commented on aten/src/ATen/gen.py:
```diff
@@ -217,13 +242,30 @@ def format_yaml(data):
+        opname = op_registration.operator_name
+        registration = op_registration.registration_code
+        # apply whitelist
+        if op_registration_whitelist and opname not in op_registration_whitelist:
```
nit: then here `if op_registration_whitelist is not None and ...`

ljk53 replied:
done

smessmer commented on aten/src/ATen/gen.py:
```diff
@@ -217,13 +242,30 @@ def format_yaml(data):
+            per_type_registrations.append(registration)
+            continue
+        # per op registration
+        per_op_registrations[opname].append(registration)
```
nit: no need for the `continue` above if we just make this the `else` case.

smessmer commented on aten/src/ATen/gen.py:
```diff
@@ -323,6 +366,10 @@ def iterate_types():
 
 
+def gen_per_op_registration_filename(opname):
+    return 'pt_op_register_{}.cpp'.format(opname.replace(':', '_'))
```
underscores are valid in op names, this could cause issues. Can we use a different character here that is valid for file names but not in op names?

ljk53 replied:
This can only cause problem when there are two op names having same filename after the replacement, e.g.: one is "aten::conv__foo" and the other is "aten__conv::foo" - feel it's pretty unlikely in reality.
Are '+' or '-' allowed in operator name? @smessmer

ljk53 replied:
changed from '_' to '-'

smessmer reviewed (approved):
looks good to me

ljk53 committed:
Update on "[pytorch] codegen flags to whitelist op registrations / generate to separate files"


This PR adds a few new parameters to ATen codegen script:

```
1. op_registration_whitelist
Can be used to filter op registrations for selective build;

2. type_whitelist
Can be used to filter types (CPUType, CUDAType, ...) for selective build;

3. per_op_registration
When set it will group function registrations by op name and write to separate files;
```

1 & 2 are introduced for mobile custom build without relying on static dispatch;
3 is introduced to solve custom build with multi-library / multi-model (needed by FB
internal build - see more details: https://fb.quip.com/ZVh1AgOKW8Vv).

These flags should work independently with each other (and independent to USE_STATIC_DISPATCH).

Not setting them should have no effect compared to master.

Differential Revision: [D19427919](https://our.internmc.facebook.com/intern/diff/D19427919/)

[ghstack-poisoned]
